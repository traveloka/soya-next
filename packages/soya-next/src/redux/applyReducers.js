import React from "react";
import hoistStatics from "hoist-non-react-statics";
import getDisplayName from "../utils/getDisplayName";
import { NEXT_STATICS } from "../constants/Statics";
import { ReactReduxContext, Provider } from "react-redux";

export default reducers => Component => {
  if (typeof reducers === "undefined") {
    return Component;
  }

  function ApplyReducers(props) {
    const flag = React.useRef(true);
    const context = React.useContext(ReactReduxContext);
    const store = props.store || context.store;

    if (flag.current) {
      if (!store.soya) {
        throw new Error(
          "applyReducers must be used with Soya's redux enhancer"
        );
      }
      store.addReducer(reducers);
      flag.current = false;
    }

    return (
      <Component {...props} />
    );
  }
  ApplyReducers.displayName = getDisplayName("applyReducers", Component);
  ApplyReducers.getInitialProps = async (ctx) => {
    if (!ctx.store.soya) {
      throw new Error(
        "applyReducers must be used with Soya's redux enhancer"
      );
    }
    ctx.store.addReducer(reducers);
    return (
      Component.getInitialProps && (await Component.getInitialProps(ctx))
    );
  }

  return hoistStatics(ApplyReducers, Component, NEXT_STATICS);
};
