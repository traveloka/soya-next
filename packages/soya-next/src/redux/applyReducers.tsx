import React from "react";
import hoistStatics from "hoist-non-react-statics";
import getDisplayName from "../utils/getDisplayName";
import { NEXT_STATICS } from "../constants/Statics";
import { ReactReduxContext } from "react-redux";

import type { AnyAction, ReducersMapObject } from "redux";
import type {
  SoyaNextPage,
  SoyaNextPageContext,
  SoyaNextStore,
} from "../types";

export interface WithApplyReducersInjectedProps {
  store?: SoyaNextStore;
}
export interface WithApplyReducersProps {
  store?: SoyaNextStore;
}

/**
 * Registers reducers to a component to enable code splitting.
 *
 * **Note**: Ensure `applyReducers([reducers])` is rendered within `createPage` hierarchy.
 * @param [reducers] An object of reducers which will be registered into a component.
 * Make sure each reducers has unique name within your application reducers.
 * @returns A higher order React component class that loads or unloads
 * registered reducers dynamically within the component lifecycle.
 */
export default function applyReducers(
  reducers?: ReducersMapObject<any, AnyAction>
) {
  return <TProps extends WithApplyReducersInjectedProps>(
    Component: SoyaNextPage<TProps>
  ) => {
    if (typeof reducers === "undefined") {
      return Component;
    }

    function ApplyReducers(
      props: Omit<TProps, keyof WithApplyReducersInjectedProps> &
        WithApplyReducersProps
    ) {
      const flag = React.useRef(true);
      const context = React.useContext(ReactReduxContext);
      const store = props.store || (context.store as SoyaNextStore);

      if (flag.current) {
        if (!store.soya) {
          throw new Error(
            "applyReducers must be used with Soya's redux enhancer"
          );
        }
        store.addReducer(reducers!);
        flag.current = false;
      }

      return <Component {...(props as TProps)} />;
    }
    ApplyReducers.displayName = getDisplayName("applyReducers", Component);
    ApplyReducers.getInitialProps = async (ctx: SoyaNextPageContext) => {
      if (!ctx.store?.soya) {
        throw new Error(
          "applyReducers must be used with Soya's redux enhancer"
        );
      }
      ctx.store.addReducer(reducers);
      return (
        Component.getInitialProps && (await Component.getInitialProps(ctx))
      );
    };

    return hoistStatics(ApplyReducers, Component, NEXT_STATICS);
  };
}
