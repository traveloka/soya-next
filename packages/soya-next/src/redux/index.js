export { default as applyReducers } from "./applyReducers";
export { default as createConfigureStore } from "./createConfigureStore";
export { default as storeEnhancer } from "./storeEnhancer";
export { default as middleware } from "./middleware";
export { default as withStore } from "./withStore";
