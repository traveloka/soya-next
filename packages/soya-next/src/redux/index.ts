export { default as applyReducers } from "./applyReducers";
export { default as createConfigureStore } from "./createConfigureStore";
export { default as storeEnhancer } from "./storeEnhancer";
export { default as middleware } from "./middleware";
export { default as withStore } from "./withStore";

export type {
  WithApplyReducersInjectedProps,
  WithApplyReducersProps,
} from "./applyReducers";
export type { WithStoreInjectedProps, WithStoreProps } from "./withStore";
