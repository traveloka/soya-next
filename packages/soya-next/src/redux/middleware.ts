import type { AnyAction, Dispatch } from "redux";

export default function middleware() {
  const queries: Record<string, any> = {};
  return (next: Dispatch<AnyAction>) =>
    ({ soya, ...action }: any) => {
      if (typeof soya !== "undefined") {
        const { load, id = action.type } = soya;
        if (typeof id !== "string") {
          throw new Error("Expected soya action id to be a string.");
        }
        if (typeof load !== "function") {
          throw new Error("Expected soya action load to be a function.");
        }

        const resolve = (soya: any) => {
          // eslint-disable-next-line callback-return
          next({
            ...action,
            soya,
          });
          delete queries[id];
          return soya;
        };
        const reject = resolve;
        if (!queries[id]) {
          queries[id] = load().then(resolve, reject);
        }
        return queries[id];
      }
      return next(action);
    };
}
