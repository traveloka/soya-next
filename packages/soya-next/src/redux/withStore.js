import React from 'react';
import PropTypes from 'prop-types';
import { storeShape } from 'react-redux/lib/utils/PropTypes';
import getDisplayName from '../utils/getDisplayName';

export default configureStore => Page => {
  const configureStoreIfNeeded = (preloadedState, extraArgument) => {
    if (typeof window === 'undefined') {
      return configureStore(preloadedState, extraArgument);
    }
    if (typeof window.store === 'undefined') {
      window.store = configureStore(preloadedState, extraArgument);
    }
    return window.store;
  };

  class WithStore extends React.Component {
    static displayName = getDisplayName('WithStore', Page);

    static propTypes = {
      reduxState: PropTypes.object.isRequired,
      store: PropTypes.oneOfType([
        PropTypes.object,
        storeShape,
      ]).isRequired,
    };

    static async getInitialProps(ctx) {
      const store = configureStoreIfNeeded(undefined, { cookies: ctx.cookies });
      const props = Page.getInitialProps && await Page.getInitialProps({ ...ctx, store });
      return {
        ...props,
        reduxState: store.getState(),
        store,
      };
    }

    constructor(props) {
      super(props);
      this.store = !process.browser ? props.store : configureStoreIfNeeded(props.reduxState, { cookies: props.cookies });
    }

    render() {
      const { ...props } = this.props;
      delete props.reduxState;
      delete props.store;

      return (
        <Page {...props} store={this.store} />
      );
    }
  }

  return WithStore;
};
