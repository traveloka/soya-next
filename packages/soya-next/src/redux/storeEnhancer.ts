import { combineReducers } from "redux";

import type { AnyAction, ReducersMapObject, StoreEnhancer } from "redux";
import type { SoyaNextStoreExt } from "../types";

export default function storeEnhancer(
  globalReducers?: ReducersMapObject<any, AnyAction>
): StoreEnhancer<SoyaNextStoreExt<any, AnyAction>, any> {
  return createStore => (reducer, preloadedState) => {
    const store = createStore(reducer, preloadedState);
    const replaceReducer = store.replaceReducer;
    let soyaReducers: ReducersMapObject<any, AnyAction> = { ...globalReducers };

    return {
      ...store,
      addReducer: nextReducers => {
        if (!nextReducers) {
          throw new Error("Missing nextReducers argument.");
        }

        if (typeof nextReducers === "object") {
          const keys = Object.keys(nextReducers);
          keys.forEach(key => {
            if (soyaReducers[key] && soyaReducers[key] !== nextReducers[key]) {
              throw new Error(`Duplicate reducer name: ${key}`);
            }
            soyaReducers[key] = nextReducers[key];
          });
          replaceReducer(combineReducers(soyaReducers));
        } else {
          replaceReducer(nextReducers);
        }
      },
      replaceReducer: nextReducers => {
        if (!nextReducers) {
          throw new Error("Missing nextReducers argument.");
        }

        if (typeof nextReducers === "object") {
          soyaReducers = {
            ...globalReducers,
            ...nextReducers,
          };
          replaceReducer(combineReducers(soyaReducers));
        } else {
          replaceReducer(nextReducers);
        }
      },
      soya: true,
    };
  };
}
