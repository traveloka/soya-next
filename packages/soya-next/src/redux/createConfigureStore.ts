import thunk from "redux-thunk";
import { applyMiddleware, combineReducers, createStore } from "redux";
import { composeWithDevTools } from "redux-devtools-extension/logOnlyInProduction";
import enhancer from "./storeEnhancer";
import middleware from "./middleware";

import type {
  Action,
  AnyAction,
  PreloadedState,
  ReducersMapObject,
} from "redux";
import type { SoyaNextStoreExt, SoyaNextThunkCompatStoreExt } from "../types";

export default function createConfigureStore(
  globalReducers?: ReducersMapObject<any, AnyAction>
) {
  return <
    TState = any,
    TAction extends Action<any> = AnyAction,
    TExtraArgs = any
  >(
    preloadedState?: PreloadedState<TState>,
    extraArgument?: TExtraArgs
  ) =>
    createStore<
      TState,
      TAction,
      SoyaNextStoreExt<TState, TAction> &
        SoyaNextThunkCompatStoreExt<TState, TAction, TExtraArgs>,
      any
    >(
      globalReducers
        ? combineReducers(globalReducers)
        : () => preloadedState || {},
      preloadedState,
      composeWithDevTools<any, any>(
        applyMiddleware(thunk.withExtraArgument(extraArgument), middleware),
        enhancer(globalReducers)
      )
    );
}
