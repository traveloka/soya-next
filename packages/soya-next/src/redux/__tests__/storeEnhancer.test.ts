import { createStore } from "redux";
import storeEnhancer from "../storeEnhancer";

import type { AnyAction } from "redux";

describe("Store Enhancer", () => {
  function setupTest() {
    return {
      store: storeEnhancer()(createStore)(state => state),
      todos: (state: any[] = [], _action: AnyAction) => state,
    };
  }

  it("should expose enhanced API", () => {
    const { store } = setupTest();
    const apis = Object.keys(store);
    expect(apis.length).toBe(7);
    expect(apis).toContain("addReducer");
    expect(apis).toContain("replaceReducer");
    expect(apis).toContain("soya");
  });

  it("should add new todos to store state", () => {
    const { store, todos } = setupTest();
    store.addReducer({ todos });
    expect(store.getState()).toMatchSnapshot();
  });

  it("should replace store start with given reducer function", () => {
    const { store, todos } = setupTest();
    const auth = (state = {}) => state;
    store.addReducer({ auth });
    store.addReducer({ todos });
    store.addReducer(todos);
    store.replaceReducer(auth);
    expect(store.getState()).toMatchSnapshot();
  });

  it("should replace store state with given reducers object", () => {
    const { store, todos } = setupTest();
    const auth = (state = {}) => state;
    store.addReducer({ auth });
    store.addReducer({ todos });
    store.replaceReducer({ auth });
    expect(store.getState()).toMatchSnapshot();
  });

  it("should throw if addReducers arguments are not specified", () => {
    expect(() => {
      const { store } = setupTest();
      // @ts-ignore for testing no arguments
      store.addReducer();
    }).toThrow("Missing nextReducers argument.");
  });

  it("should throw if replaceReducers arguments are not specified", () => {
    expect(() => {
      const { store } = setupTest();
      // @ts-ignore for testing no arguments
      store.replaceReducer();
    }).toThrow("Missing nextReducers argument.");
  });

  it("should throw if there's reducer name conflict", () => {
    expect(() => {
      const { store, todos } = setupTest();
      store.addReducer({ todos });
      store.addReducer({ todos: (state: any) => state });
    }).toThrow("Duplicate reducer name: todos");
  });
});
