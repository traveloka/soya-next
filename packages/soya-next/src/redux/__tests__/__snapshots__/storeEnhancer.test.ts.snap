// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Store Enhancer should add new todos to store state 1`] = `
Object {
  "todos": Array [],
}
`;

exports[`Store Enhancer should replace store start with given reducer function 1`] = `
Object {
  "auth": Object {},
  "todos": Array [],
}
`;

exports[`Store Enhancer should replace store state with given reducers object 1`] = `
Object {
  "auth": Object {},
}
`;
